src/token.rs

```diff
@@ -413,23 +312,62 @@ impl Token {
         }
     }
 
+    fn clear_private_session_objects(&mut self) {
+        let mut priv_uids = Vec::<String>::new();
+        for obj in &self.storage.get_all_cached() {
+            if obj.is_private() {
+                let oh = obj.get_handle();
+                if oh != CK_INVALID_HANDLE {
+                    let _ = self.handles.remove(oh);
+                }
+                if !obj.is_token() {
+                    /* not a token object, therefore we need to destroy it */
+                    if let Ok(uid) = obj.get_attr_as_string(CKA_UNIQUE_ID) {
+                        priv_uids.push(uid.clone());
+                    }
+                }
+            }
+        }
+
+        /* remove all private session objects */
+        for uid in priv_uids {
+            let _ = self.storage.remove_by_uid(&uid);
+        }
+    }
+
+    fn clear_session_objects(&mut self, handle: CK_SESSION_HANDLE) {
+        let mut handles: Vec<CK_OBJECT_HANDLE> = Vec::new();
+        for obj in &self.storage.get_all_cached() {
+            if obj.get_session() == handle {
+                if !obj.is_token() {
+                    handles.push(obj.get_handle());
+                }
+            }
+        }
+
+        for h in handles {
+            if let Some(uid) = self.handles.remove(h) {
+                let _ = self.storage.remove_by_uid(&uid);
+            }
+        }
+    }
+
     pub fn get_object_by_handle(
-        &self,
-        handle: CK_OBJECT_HANDLE,
-        checks: bool,
+        &mut self,
+        o_handle: CK_OBJECT_HANDLE,
     ) -> KResult<&Object> {
-        let obj = match self.objects.get_by_handle(handle) {
-            Ok(o) => o,
-            Err(e) => return Err(e),
+        let is_logged_in = self.is_logged_in(KRY_UNSPEC);
+        let obj = match self.handles.get(o_handle) {
+            Some(s) => self.storage.fetch_by_uid(s)?,
+            None => return err_rv!(CKR_OBJECT_HANDLE_INVALID),
         };
-        if checks && !self.is_logged_in(KRY_UNSPEC) && obj.is_private() {
+        if !is_logged_in && obj.is_private() {
             return err_rv!(CKR_USER_NOT_LOGGED_IN);
         }
         Ok(obj)
     }
 
     fn validate_pin_obj(
-        &self,
         obj: &Object,
         label: String,
     ) -> KResult<(Vec<u8>, CK_ULONG)> {
```

Comment:
I do not think they can be propagated down because the UNIQUE_ID is always forcibly generated and is a UUID, therefore it can never produce the speaicl UNIQUE_ID value that identify these (on purpose).

But we should definitely filter them, they should not be retrievable by applications.

Reply:
Filled #23.

Reply:
Oh, I see you already did this in this PR. So lets close it when we will merge this one (and after we will verify its fixed).